File: sdk/python/altwallet_sdk/exceptions.py
"""Custom exceptions for the AltWallet Python SDK."""

from typing import Any


class AltWalletError(Exception):
    """Base exception for all AltWallet SDK errors."""

    def __init__(
        self,
        message: str,
        error_code: str | None = None,
        request_id: str | None = None,
        details: dict[str, Any] | None = None,
    ):
        super().__init__(message)
        self.message = message
        self.error_code = error_code
        self.request_id = request_id
        self.details = details or {}

    def __str__(self) -> str:
        parts = [self.message]
        if self.error_code:
            parts.append(f" (Code: {self.error_code})")
        if self.request_id:
            parts.append(f" (Request ID: {self.request_id})")
        return " ".join(parts)


class ConfigurationError(AltWalletError):
    """Raised when SDK configuration is invalid."""

    pass


class NetworkError(AltWalletError):
    """Raised when network communication fails."""

    pass


class AuthenticationError(AltWalletError):
    """Raised when authentication fails."""

    pass


class ValidationError(AltWalletError):
    """Raised when request validation fails."""

    pass


class RateLimitError(AltWalletError):
    """Raised when rate limit is exceeded."""

    pass


class APIError(AltWalletError):
    """Raised when the API returns an error response."""

    def __init__(
        self,
        message: str,
        status_code: int,
        error_code: str | None = None,
        request_id: str | None = None,
        details: dict[str, Any] | None = None,
    ):
        super().__init__(message, error_code, request_id, details)
        self.status_code = status_code

    def __str__(self) -> str:
        return f"{super().__str__()} (Status: {self.status_code})"

File: sdk/nodejs/src/exceptions.ts
/**
 * Custom exceptions for the AltWallet Node.js SDK.
 */

export class AltWalletError extends Error {
  public readonly errorCode?: string;
  public readonly requestId?: string;
  public readonly details?: Record<string, any>;

  constructor(
    message: string,
    errorCode?: string,
    requestId?: string,
    details?: Record<string, any>
  ) {
    super(message);
    this.name = 'AltWalletError';
    this.errorCode = errorCode;
    this.requestId = requestId;
    this.details = details || {};

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, AltWalletError);
    }
  }

  toString(): string {
    const parts = [this.message];
    if (this.errorCode) {
      parts.push(`(Code: ${this.errorCode})`);
    }
    if (this.requestId) {
      parts.push(`(Request ID: ${this.requestId})`);
    }
    return parts.join(' ');
  }
}

export class ConfigurationError extends AltWalletError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 'CONFIGURATION_ERROR', undefined, details);
    this.name = 'ConfigurationError';
  }
}

export class NetworkError extends AltWalletError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 'NETWORK_ERROR', undefined, details);
    this.name = 'NetworkError';
  }
}

export class AuthenticationError extends AltWalletError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 'AUTHENTICATION_ERROR', undefined, details);
    this.name = 'AuthenticationError';
  }
}

export class ValidationError extends AltWalletError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 'VALIDATION_ERROR', undefined, details);
    this.name = 'ValidationError';
  }
}

export class RateLimitError extends AltWalletError {
  constructor(message: string, details?: Record<string, any>) {
    super(message, 'RATE_LIMIT_ERROR', undefined, details);
    this.name = 'RateLimitError';
  }
}

export class APIError extends AltWalletError {
  public readonly statusCode: number;

  constructor(
    message: string,
    statusCode: number,
    errorCode?: string,
    requestId?: string,
    details?: Record<string, any>
  ) {
    super(message, errorCode, requestId, details);
    this.name = 'APIError';
    this.statusCode = statusCode;
  }

  toString(): string {
    return `${super.toString()} (Status: ${this.statusCode})`;
  }
}
